Implement optional hashing of tokens and secrets

oliverguenther opened the pull request:
There have been previous requests with regards to encryption or hashing OAuth secrets in the database to avoid storing them in plain texts. Primarily, these discussions have taken place in #320 and #675 (at
least, that's where I took note of them).

For us, adaptation of doorkeeper is limited due to this exact storing of plain text tokens, especially longlived refresh tokens as well as client secrets in the database. Different from some of the commenters in the above discussion, since tokens are already cryptographically secure randoms, a secure hash function is sufficient to address the immediate threat of leaking tokens and secrets, e.g., from an unencrypted backup. 

Encryption does not provide much of an advantage IF an attacker is able to get access to the resource server and leak the encryption key. (granted, it's a big IF and strong adversary, but still)

This PR thus implements an *optional* hashing filter configurable
through the initializer that:

1. Hashes access_tokens, refresh_tokens, access_grants and application
client_secrets.

2. Allow models to maintain the volatile plaintext value for one-time communication to the user (e.g., showing the application, token in the same request).

3. Alters lookups to allow finding input tokens through either their hashed value, or the original plaintext as fallback. This allows users to switch on this hashing option without losing any old authenticated clients. They may want to reduce TTL of their refresh tokens to ensure all clients will have hashed secrets persisted in the near future.

Arguments against some of the discussions:

1. Using bcrypt. Since the original plaintext token is a cryptographically secure random string, we do not gain much by salting the values. The UniqueToken generator of doorkeeper has 128-bit entropy by default, and can be extended. However, this implementation allows users to use bcrypt or
whatever they prefer, it is just my opinion that the extra overhead is useless.

2. Not storing tokens at all whereever possible. This was a fascinating read, but out of my scope of what I'm capable to implement and would be completely orthogonal to what doorkeeper currently does. I for one like having the tokens around for auditing purposes.

I have deliberately not updated neither news nor contributions yet while we're still discussion what this PR will lead to.

nbulaj commented on lib/doorkeeper/config.rb:
```diff
@@ -241,6 +241,11 @@ def option(name, options = {})
+    # Allow optional hashing of input tokens to some other representation
+    # that is persisted. Can be used for hashing of input token,
+    # grants and application secret values.
+    option :secret_hashing, default: nil
```
I prefer (and think it's something like an enterprise standard)  `encrypt_secrets` instead of `hashing`, WDYT?

oliverguenther replied:
It's only terminology and I will accept your decision, but for me `encrypt_secrets` bears no resemblance to what this option will do. It will irreversably map all plain text tokens with a one-way function, so encryption does not really fit the option in my opinion.

nbulaj replied:
OK, now I've get the idea. Let's rename it to `hash_secrets`, WDYT (like action, `reuse token`, `enable_application_owner` , `enforce_content_type`, etc)?

oliverguenther replied:
Aye! :+1:

nbulaj commented on lib/doorkeeper/config.rb:
```diff
@@ -286,10 +291,15 @@ def option(name, options = {})
+    ##
+    # Reuse of access tokens is only available when tokens are not hashed
+    def reuse_access_token
+      @reuse_access_token unless hash_secrets?
```
I think it would be better to log some message like `"You are activated reuse access token with secrets encryptions that is unsupported. This option will be ignored"` instead of just silently return `false`

oliverguenther replied:
I fully agree on that one

nbulaj commented on lib/doorkeeper/config.rb:
```diff
@@ -322,6 +332,18 @@ def raise_on_errors?
       handle_auth_errors == :raise
     end
 
+    def hash_secrets?
```
The same - `encrypt_secrets?`

nbulaj commented on lib/doorkeeper/config.rb:
```diff
@@ -322,6 +332,18 @@ def raise_on_errors?
+      !@secret_hashing.nil?
+    end
+
+    def optional_hash(plain_token)
```
This naming is not fully clear to me.. Maybe we can rename it to something more transparent? I don't know, like `encrypted_or_plain_token(token)` (if I understood correctly the behavior of this method)

oliverguenther replied:
The method will return the hashed plaintext token value IFF token hashing is enabled, otherwise it will return the plain value.

In the same spirit as the other comments, I would suggest we stick to the name `hashing` (vs encryption). What about `hashed_or_plain_token(plain_token)`  then?

nbulaj replied:
Ok, so we'' work with hashes. Yes, `hashed_or_plain_token` would be better :+1:

nbulaj commented on lib/doorkeeper/models/access_grant_mixin.rb:
```diff
@@ -30,7 +31,7 @@ module ClassMethods
       #
       def by_token(token)
-        find_by(token: token.to_s)
+        find_by_optional_hash(:token, token)
```
The same here (about naming)..

nbulaj commented on lib/doorkeeper/models/concerns/hashable.rb:
```diff
@@ -0,0 +1,37 @@
+    ##
+    # Hashable finder to provide lookups for input plaintext values which are
+    # mapped to their hashes before lookup.
+    module Hashable
```
I prefer `Encryptable` over `Hashable`. However, RFC 6819 allows both - encrypt / hash credentials (check out 5.1.4.1.4.  Encryption of Credentials for example).

BTW, very good point to extract this functionality to separate concern that could be reused in any Doorkeeper gem extension 👍

oliverguenther replied:
Again, as above, the option added is `one-way` and thus will not allow for encryption of token values. It's possible to alter that, but with my rationale in the opening statement, we should avoid encryption in all cases where authorization flow does not require it. Since doorkeeper implements no such scheme where e.g., the `client_secret` needs to be accessed in clear text, there is no advantage of encryption over hashing.

nbulaj replied:
Yep, now I understand. Agree

nbulaj commented on lib/doorkeeper/models/concerns/hashable.rb:
```diff
@@ -0,0 +1,37 @@
+        # @return [Doorkeeper::AccessToken, nil] AccessToken object or nil
+        #   if there is no record with such token
+        #
+        def find_by_optional_hash(attr, token)
```
I think we can hide from the end-user our internal API and call this method something like `find_by_token` and do all the magic (use plain token or encrypted one if encryption activated) in it

oliverguenther replied:
There is already `by_token` in the AccessGrant and AccessToken mixins that are meant as the public interface. These are user facing, so my suggestion would simply be to mark this method `:private` to make it clear it is not a public interface.

nbulaj replied:
It's a good idea to mark it as a private (internal) API, but it's opened for developers so anyone could easily use it.. I'll reconsider it in a meanwhile

Also I would still think about the naming..

oliverguenther replied:
I believe `find_by_hashed_or_plain_token`, while a tad long, describes exactly what it does. It's identical to what the helper does. If token hashing is active, it will look for a hashed token, or fall back to looking up a plaintext token.

nbulaj replied:
It's great :+1:

oliverguenther commented:
The last commit addresses:


- Rename `secret_hashing` option to `hash_secrets`
- Add warning when `reuse_access_token` is enabled together with
`hash_secrets`, but ensure it's printed only once during initialization
- Renamed `optional_hash` helper to `hashed_or_plain_token`
- Renamed `find_by_optional_hash` to `find_by_hashed_or_plain_token`
- Made `find_by_hashed_or_plain_token` private

This now also introduces a very basic `validate` method in the configuration to ensure the warning printed when `reuse_access_token` and `hash_secrets` are both active will be printed only once during configuration. Putting the warning in the getter will flood the user log.

nbulaj commented on lib/doorkeeper/config.rb:
```diff
@@ -46,7 +46,7 @@ def initialize(&block)
 
       def build
-        @config
+        @config.valid_subset
```
I personally prefer something like just `validate`, or `validate!`

oliverguenther replied:
That's fine with me, changed to `validate`

nbulaj commented on lib/doorkeeper/config.rb:
```diff
@@ -392,5 +393,19 @@ def calculate_token_grant_types
+    # have both been activated.
+    #
+    # In that case, disable reuse_access_token value and warn the user.
+    def validate_reuse_access_token_value
```
Maybe we can move it in `reuse_access_token` setter ? It works once (on config build), so the behavior will be the same, but we wouldn't change the Configuration module API (no need to change `build` method)

oliverguenther replied:
I initially tried that, but it will not work if you place `hash_secrets` after `reuse_access_token`. This will require the check at two places and that's why I moved it out of there. I can still move it in both setters if you prefer.

nbulaj commented:
Also could you please add a changelog entry as gem bot suggested? :)

oliverguenther commented:
> Also could you please add a changelog entry as gem bot suggested? :)

Added, also squashed all commits from the first round now.

nbulaj commented:
@f3ndot could you please assist with this PR ? Maybe you as a security expert find something I missed :)

oliverguenther commented on lib/doorkeeper/orm/active_record/application.rb:
```diff
@@ -45,14 +45,29 @@ def self.revoke_tokens_and_grants_for(id, resource_owner)
-      self.secret = UniqueToken.generate if secret.blank?
+      return unless secret.blank?
+
+      @volatile_secret = UniqueToken.generate
```
Sure, no problem

f3ndot commented:
Another question: Should Doorkeeper be clever enough to (1) if hashing is enabled and (2) plaintext fallback occurs that it silently hashes the value and persists to the database?

Generally side-effects are a no-no but it might be a nice migration path for free for users of the gem

f3ndot commented on lib/doorkeeper/models/application_mixin.rb:
```diff
@@ -49,5 +49,14 @@ def by_uid(uid)
+    # To fall back to older plaintext applications, also compares plain values.
+    def secret_matches?(input)
+      secret == Doorkeeper.configuration.hashed_or_plain_token(input) ||
+        secret == input
```
I realize now that if we make `hash_application_secrets` configurable you'll need to expose an option block for handling verification:

```ruby
verify_application_secrets do |ciphertext, input|
  BCrypt::Password.new(ciphertext) == input
end
```

```ruby
def secret_matches?(input)
  verify_application_secrets.call(secret, input) || secret == input
end
```

oliverguenther replied:
> Technically this could cause a problem where the hashed secret of one application collides with the plaintext of another presenting a security issue, but the odds of that are so ridiculously low it's barely worth mentioning.

Indeed. I suggest we add a note to the configuration that a cryptographically secure, collision-resistant one-way function is mandatory. I did not want to coerce a specific digest mechanism.

oliverguenther replied:
I can follow your argument regarding application secrets, and while bruteforcing even unsalted SHA256 is and will remain unfeasible for quite some time, I don't mind using a split method for either.

Maybe we should not expose the actual hashing function but simply choose SHA256 and BCrypt ? It would reduce the error potential in users to choose a wrong or insecure method or implement it wrongly. /cc @nbulaj  what do you think?

f3ndot replied:
There's an expression: _"Give 'em enough rope, and they'll hang themselves"_

I'm comfortable with making Doorkeeper decide these secure hashing algorithms instead of leaving it to a developer who, while well intentioned, makes a critical error.

We can always consider introducing configuration if people raise an issue in the future.

nbulaj replied:
I think it would be great to propose a more secured and stable solution by default, but definetly with the ability to reconfigure it (for those developers that know what they are doing). WDYT?

f3ndot replied:
From a security standpoint secure and sane defaults is just as good as no configuration at all. I leave it up to y'all to decide what's right for Doorkeeper.

A good comment explaining rationale for why bcrypt and sha256 goes a long way should we opt for exposing configuration.

nbulaj replied:
I had a lot of cases (both my own and my colleagues too) when we need to patch default gem behavior just because it couldn't be configured and doesn't suite our (customer) needs (maybe not connected with hashing algorithms, but ...), so I prefer to have default behavior **with the ability** to change it (instead of monkey-patching :see_no_evil:  ). Also in such case everybody could migrate to any hashing solution in case when Bcrypt (%`your_solution`%) would be easy to brute-force (without need to update to a new Doorkeeper release especially if it breaks existing API).

oliverguenther replied:
The problem with the configuration option is that we need that twice if we're going for different variants for application and token secrets.

My suggestion is we have a setter on the `Hashable` mixin that allows implementing classes to define a secret hashing and verification function. It will use `SHA256.hexdigest` by default and will be overridden by `Application` to use BCrypt.

In the same fashion, an experienced user can override these sensible defaults by e.g., using refinements or simply by setting 

```
      self.secret_hash_function = lambda do |plain_token|
        BCrypt::Password.create(plain_token.to_s)
      end
```

without the need to check for any other configuration options (such as whether hashing is enabled for this class). My code will respect that defined method regardless whether it comes from the class or another user.

It shouldn't be too easy to override these methods because:

1. The user must be absolutely certain their implementation is correct
2. They must never change this hashing method without additional migration efforts. Putting a configuration option **suggests** that it can be turned on and off as they like.

oliverguenther commented on lib/doorkeeper/models/concerns/hashable.rb:
```diff
@@ -0,0 +1,41 @@
+        # @return [Doorkeeper::AccessToken, nil] AccessToken object or nil
+        #   if there is no record with such token
+        #
+        def find_by_hashed_or_plain_token(attr, token)
```
That is absolutely fine with me. 👍

oliverguenther replied:
Looking at the code again, your helpers already exist at the model level. I believe the `Hashable` concern is correct and should not know details about what attributes exist (Otherwise you'll have `find_by_refresh_token` in the concern resulting in it existing in the access grant model which doesn't know refresh tokens.

The places you're looking for are those two, they should be extended exactly the same like in your current setup.
https://github.com/doorkeeper-gem/doorkeeper/pull/1168/files#diff-dc86e93a23661f9a5d87d2b031af8ba5R34
https://github.com/doorkeeper-gem/doorkeeper/pull/1168/files#diff-c42935829979fd1bcb5ed418c92f6386R26

oliverguenther commented:
> Another question: Should Doorkeeper be clever enough to (1) if hashing is enabled and (2) plaintext fallback occurs that it silently hashes the value and persists to the database?

The decision to fallback is by design, otherwise existing doorkeeper implementations could never make use or transition to hashed tokens, and I argue this kind of behavior is industry standard for upgrading e.g., older secret storage implementations without interrupting user flow. It is still up to the application to discard all old tokens to ensure no-one can access with plaintext tokens.

I would respect your opinion and add an off-by-default flag that allows plaintext fallback, since I believe it would benefit a lot of users and in the end, improve practical security of Doorkeeper, since users will not be able to enable this option in a future release.

f3ndot commented:
@oliverguenther I don't think I explained myself correctly. When someone upgrades Doorkeeper to the version that hashes by default and they have a bunch of plaintext tokens, instead of expecting them to manually figure out how to backfill the plaintext or leaving those entries as plaintext forever, we make a code change that in fallback mode there's a side effect of hashing the plaintext value and saving it.

I don't think it's the right thing to do but was worth mentioning aloud nonetheless.

oliverguenther commented:
@f3ndot I think we understood each other correctly. My suggestion is that I add an explicit flag `fallback_to_plain_secrets` that allows those users switching on hashing on existing installations to ensure plain values are being found.

This will not be a side-effect, but a deliberate option users receive to make the transition. New users will never receive plain token lookups, and thus the (ever so theoretical) option of matching a collision does not exist from them.

oliverguenther commented:
@f3ndot @nbulaj I just pushed a new commit (deliberately not squashed to show the changes) that

1. Adds separate configuration options 
  - `hash_token_secrets` To control AccessToken, AccessGrant hashing
  - `hash_application_secrets` To control Application secret hashing
  - `fallback_to_plain_secrets` To control automatical fallback + upgrading of plain text token values (see below)

The hashing function and comparer is not configurable in the user-facing configuration, but could be overridden for each including class (and is for Application) by using `Model.secret_hash_function=` as can be seen here https://github.com/doorkeeper-gem/doorkeeper/pull/1168/commits/4684568ce93a238169c7bcd909af9d982d2568a4#diff-d0fc797f13ea4de1171e4f62a72a28e0R17 . This implements BCrypt for Application secret hashing while all reusing a single `Hashable` interface to avoid additional complexity and exposing 6 configuration pairs for all secret-holding classes (one each for hashing function and comparer).

This will make it dead-easy for the 99th percentile to switch on secret hashing by updating Doorkeeper and uncommenting all three boolean options. They will not have to think about secure hashing algorithms or migration of their plaintext values.

However, I want to note the danger of the `fallback_to_plain_secrets` side-effect option: If application code EVER were to use `AccessToken.by_token("the hashed token here")`, it would get rehashed because we cannot distinguish an old plaintext token from its SHA256 digest. My suggestion would thus be either:

1. We must make the interaces `by_token` and `by_refresh_token` clear that these must never receive the hashed tokens. There is nothing to prevent this though, without an additional column.
2. We should instead remove that option of automatically upgrading and create a rake task that users have to execute when enabling that option.

nbulaj commented:
Hi @oliverguenther . I like the proposed solution.

What about fallback - all I want is the:

1) backward compatibility
2) easy of usage

So all the developers that will update their dependencies to Doorkeeper 5.1 (for example) don't need to change anything if they don't need to hash secrets. 

New projects can simply decide what they want to use - plain or hashed secrets.

But what about legacy projects? What if they need to enable hashing option ? I think we need to propose a simple API for our models / classes / concerns and cover migration process (both in the code and Wiki page). WDYT ?

oliverguenther commented:
@nbulaj  From an application developer simply using Doorkeeper, all your points are already available for them:

1. Upon upgrading without any configuration changes, there is nothing to take care of, Doorkeeper will continue without hashing as before.
2. Whenever a user decides to enable `hash_token_secrets` or `hash_application_secrets`, they can enable `fallback_to_plain_secrets` as documented in the initializer to allow any tokens from their previous version to continue to work (and as @f3ndot requested, automatically converted to hashed tokens).

Only when:

-  the user **does not enable** `fallback_to_plain_secrets`, it will disregard all previous plaintext secrets and assume they are invalid. I believe this as well is a valid option to go to if the user decides not to allow any previous plaintext secrets.

- A developer **modifying** Doorkeeper must take care not to enable `fallback_to_plain_secrets` and passing hashed values into the finder methods. This is simply due to the side-effect of hashing plaintext values.

oliverguenther commented:
Only saw your message now @nbulaj , thanks for all the feedback. @f3ndot Do you have further remarks especailly regarding the passage above?

nbulaj commented:
Hi all. I'm planning to merge it and release after the holidays, so we have some time if this PR requires some changes. So just pinging @f3ndot

oliverguenther commented:
Happy new year! I have force-pushed this into a single commit again to resolve the conflict. If @f3ndot still needs the logical second commit after the review feedback, I've created a temporary branch for it here: https://github.com/doorkeeper-gem/doorkeeper/compare/master...oliverguenther:feature/hashed-tokens-nonrebased?expand=1

nbulaj commented:
Hi @oliverguenther . Let's update it again and I'll merge this changes. Don't look at Travis because we face some issues with Bundler 2.0 and ruby on Travis :)

oliverguenther commented:
Alright, thanks! I've just rebased on master again.

oliverguenther commented:
And rebased again due to conflicts in `NEWS.md`.

nbulaj commented:
Big thanks @oliverguenther for your patience and contribution! 
I think it is a good start for something extra great for Doorkeeper project :)

oliverguenther commented:
Thanks for your efforts in the gem and this PR! :beers:

f3ndot commented:
Thank you both for this! @oliverguenther and @nbulaj 

Apologies I fell off the face of the earth there.
